Goroutine leak in TestEtcdGrpcResolver

I tried to figure out if there might be goroutine leaks in unit tests. So I used [goleak](https://github.com/uber-go/goleak) to guard all test functions (`main` branch) and run `PASSES=unit ./test`. I finally found a test leaking goroutines:

```
--- FAIL: TestEtcdGrpcResolver (4.43s)
    resolver_test.go:51: Cannot create clusters in --short tests
     leak.go:117: Test appears to have leaked :
         google.golang.org/grpc.(*Server).Serve.func1(0xc00042ee00)
         	/root/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:715 +0xe5
         google.golang.org/grpc.(*Server).Serve(0xc00042ee00, 0x18485f0, 0xc0004a11a0, 0x0, 0x0)
         	/root/go/pkg/mod/google.golang.org/grpc@v1.37.0/server.go:769 +0x88a
                created by go.etcd.io/etcd/pkg/v3/grpc_testing.(*StubServer).Start
        	/root/actions-runner/_work/path/to/etcd/pkg/grpc_testing/stub_server.go:53 
FAIL
FAIL	go.etcd.io/etcd/tests/v3/integration/clientv3/naming	39.512s
```

In normal situations, there should be no alive goroutines after a test function  finished. Is this goroutine leak intentional or expected?

--- a/pkg/grpc_testing/stub_server.go
+++ b/pkg/grpc_testing/stub_server.go
@@ -26,10 +26,14 @@ type StubServer struct {
 	s *grpc.Server
 
 	cleanups []func() // Lambdas executed in Stop(); populated by Start().
+	started chan struct{}
 }
 
 func New(testService testpb.TestServiceServer) *StubServer {
-	return &StubServer{testService: testService}
+	return &StubServer{
+		testService: testService,
+		started: make(chan struct{}),
+	}
 }
 
 // Start starts the server and creates a client connected to it.
@@ -50,7 +54,10 @@ func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption)
 
 	s := grpc.NewServer(sopts...)
 	testpb.RegisterTestServiceServer(s, ss.testService)
-	go s.Serve(lis)
+	go func() {
+		close(ss.started)
+		s.Serve(lis)
+	}()
 	ss.cleanups = append(ss.cleanups, s.Stop)
 	ss.s = s
 
@@ -59,6 +66,7 @@ func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption)
 
 // Stop stops ss and cleans up all resources it consumed.
 func (ss *StubServer) Stop() {
+	<-ss.started
 	for i := len(ss.cleanups) - 1; i >= 0; i-- {
 		ss.cleanups[i]()
 	}

--- a/pkg/grpc_testing/stub_server.go
+++ b/pkg/grpc_testing/stub_server.go
@@ -67,7 +67,6 @@ func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption)
 // Stop stops ss and cleans up all resources it consumed.
 func (ss *StubServer) Stop() {
 	<-ss.started
-
 	for i := len(ss.cleanups) - 1; i >= 0; i-- {
 		ss.cleanups[i]()
 	}

--- a/pkg/grpc_testing/stub_server.go
+++ b/pkg/grpc_testing/stub_server.go
@@ -26,13 +26,13 @@ type StubServer struct {
 	s *grpc.Server
 
 	cleanups []func() // Lambdas executed in Stop(); populated by Start().
-	started chan struct{}
+	started  chan struct{}
 }
 
 func New(testService testpb.TestServiceServer) *StubServer {
 	return &StubServer{
 		testService: testService,
-		started: make(chan struct{}),
+		started:     make(chan struct{}),
 	}
 }
 
@@ -67,6 +67,7 @@ func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption)
 // Stop stops ss and cleans up all resources it consumed.
 func (ss *StubServer) Stop() {
 	<-ss.started
+
 	for i := len(ss.cleanups) - 1; i >= 0; i-- {
 		ss.cleanups[i]()
 	}

